DOC: enhance documentation for first-time users

stsievert opened the pull request:
**What does this PR implement?**
It adds a link back to the HTCondor homepage on the front page of the HTMap documentation.

I was looking for a piece of documentation specific to HTCondor ("what does a HOLD tag on a job mean?"). I went looking for link back to the HTCondor documentation, and the front page was the first place I looked.

It also implements the following:

* Adds page on the CLI
* Adds page on how to effectively use HTCondor (requesting CPUs/disk/GPUs/etc, some useful shell commands, etc)
* Reorganizes documentation (rename "recipes" to "advanced tutorials")
* Adds example about processing results from completed job to `Map.component_by_status`
* Adds note about how string parameters are processed (including note about strings with quotes)

All of these come from my first (real) usage of HTMap and HTCondor.

*edit* This PR will...

* ~~close #201~~
* close #206

JoshKarpel commented on docs/source/faq.rst:
```diff
@@ -24,3 +24,6 @@ See the "Attention" box near the top of :doc:`dependencies`.
 
+I'm getting an error about a job being held
+-------------------------------------------
+Your code likely through an error. See :doc:`tutorials` for more detail.
```
```suggestion
I'm getting an error about a job being held. What should I do?
--------------------------------------------------------------------
Your code likely encountered an error during remote execution. See the :doc:`tutorials` for more details on how to troubleshoot remove execution errors.
```

stsievert commented:
Let's hold off a little bit on merging; I'd like to at least get a basic test up and running. This PR can hold the documentation fixes I make in the meantime.

I have perhaps a naive usage question. I have this script on `submit2.chtc.wisc.edu`:

``` python
def job(x):
    return x

if __name__ == "__main__":
    import htmap as ht

    future = ht.map(job, range(4))
    print(future)
    results = list(future)
    print(results)

```

I submit this script with my personal miniconda install. It throws an error because `ExitCode =!= 0`. Here's stdout with the traceback:

```
(base) [stsievert@submit2 htmap-trial]$ python submit.py
Map(tag = super-proper-rig)
Traceback (most recent call last):
  File "submit.py", line 11, in <module>
    results = list(future)
  File "/home/stsievert/miniconda3/lib/python3.7/site-packages/htmap/maps.py", line 422, in __iter__
    yield from self.iter()
  File "/home/stsievert/miniconda3/lib/python3.7/site-packages/htmap/maps.py", line 439, in iter
    yield self._load_output(component, timeout = timeout)
  File "/home/stsievert/miniconda3/lib/python3.7/site-packages/htmap/maps.py", line 343, in _load_output
    self._wait_for_component(component, timeout)
  File "/home/stsievert/miniconda3/lib/python3.7/site-packages/htmap/maps.py", line 309, in _wait_for_component
    raise exceptions.MapComponentHeld(f'Component {component} of map {self.tag} is held: {self.holds[component]}')
htmap.exceptions.MapComponentHeld: Component 0 of map super-proper-rig is held: [3] The job attribute OnExitHold expression 'ExitCode =!= 0' evaluated to TRUE
(base) [stsievert@submit2 htmap-trial]$
```

Why isn't this working?

JoshKarpel commented:
I snooped around in your `.htmap` directory on submit2 (hope you don't mind), since that error is a little odd.

It turns out you're falling through a crack in the mechanism that detects execution errors. The problem is that your Docker image doesn't have `cloudpickle` installed. If you run `future.stderr[0]` you can see the actual error; if you run `future.stdout[0]` you'll see the the installed packages.

Basically, the problem is that HTMap is failing execute-side before it can manage to produce an execution error for you. I'm not quite sure how to tackle this, but I've created an issue to track it https://github.com/htcondor/htmap/issues/194.

For now, you probably need to make a new image (based on the one you're using) with `cloudpickle` (or preferably `htmap` itself) installed.

stsievert commented:
> I snooped around in your .htmap directory on submit2 (hope you don't mind), since that error is a little odd.

Thank you – that resolved my issue. I've added some debugging tips in `error-handling.ipynb` for when a job is held. Now, I have the following to complete:

1. Launching with Docker
2. Machine specification to run jobs
3. Running my job
4. (possibly) checkpointing.

I'll likely have more documentation changes in this process. I've modified PR title to reflect that.

stsievert commented:
I've consolidates the tutorials and recipes into one page with a divider. I can't test this change right now (htcondor doesn't install on macOS).

I did this because I remembered seeing a Docker tutorial, and spent a couple minutes looking for it. I first looked in "Tutorials", then I looked in "Dependency Management", then I bounced around a couple places before looking in the recipes. I moved it to the first place I looked.

JoshKarpel commented:
Seems quite reasonable!

If you need an environment to build the docs in, the testing Docker container should just need the docs requirements installed in it to be able to do a docs build. Start the container by running `./dr bash` from the top of the repository, `pip install -r docs/requirements.txt`, and then you should be able to `cd docs` and `make html`.

I should probably encapsulate that in another container for ease of use... (I run into the same problem on Windows, since we don't have pip-installable HTCondor Python bindings on Windows).

JoshKarpel commented:
Quick thought - be careful on HTCondor documentation, because a lot of things ends up being site-specific. The `WantFlocking` option, for example, is CHTC-only. We need to avoid CHTC-isms in the docs.

That does make me think that we should eventually have a "HTMap in CHTC" page on https://github.com/CHTC/chtc-website-source where we can put that kind of information.

JoshKarpel commented:
@stsievert been a while since you committed on this - think it's ready for review & merge?

stsievert commented:
Probably. I've added all I'm going to add. I think there's some reorganization to do still in the sidebar and in the titles. Should I put that in this PR, another PR or do you want to handle it?

JoshKarpel commented:
I say go ahead and do it in this PR if you think you can get it done in the next week or two (I'd like to release v0.6.0 with these improvements included soon). Otherwise, I'd be happy to try to the reorg as you recommend, or defer it to another PR.

stsievert commented on docs/source/faq.rst:
```diff
@@ -3,18 +3,36 @@ FAQ
+`HTCondorCluster`_. After `HTCondorCluster`_ is used, Dask can be used as
+normal or on your own machine. This is common with other cluster managers too:
+Dask-Jobqueue also wraps SLURM, SGE, PBS and LSF clusters, and Dask Distributed
+can wrap Kubernetes and Hadoop clusters. This interface will not use HTMap.
```
There are no downsides mentioned in use of `HTCondorCluster`. Should some be added?

stsievert replied:
Thanks. I'll let you do that – you'll write a better reason than me.

stsievert commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+.. _cli-tips:
+
+Separate job submission/monitoring/collection
+---------------------------------------------
```
I'm not sure this section is accurate and that it's actually a tip/trick. Can you review this section?

JoshKarpel commented on docs/source/htcondor.rst:
```diff
@@ -0,0 +1,100 @@
+   )
+   ht.map(..., map_options=options)
+
+When it's mentioned that "the option ``foo`` needs to be set" (possibly in a
+submit file), this corresponds to adding the option in the appropriate
+place in :class:`~MapOptions` (see the documentation for details).
```
This link didn't work; I suspect this file is missing a `.. py:currentmodule:: htmap` at the top

JoshKarpel commented on docs/source/htcondor.rst:
```diff
@@ -0,0 +1,100 @@
+Shell commands
+--------------
+
+Here are some shell HTCondor commands and their primary use:
```
```suggestion
HTMap tries to expose a complete interface for submitting and managing jobs,
but not for examining the state of your HTCondor pool itself.
Here are some HTCondor shell commands that you may find useful:
```

JoshKarpel commented on docs/source/htcondor.rst:
```diff
@@ -0,0 +1,100 @@
+
+.. code:: shell
+
+   ## See the jobs you've submitted, and refresh them every 2 seconds
+   watch condor_q --submitter foobar
+
```
This one is heretical; the facilitators yell when they see people doing it :)

```suggestion
```

stsievert replied:
...I don't know why I submitted that?

🤥

stsievert replied:
I've removed the `watch`, so now it's just `condor_q --submitter foobar`.

JoshKarpel commented on docs/source/htcondor.rst:
```diff
@@ -0,0 +1,100 @@
+
+   ## See how much CUDA memory on each machine (and how many are available)
+   condor_status --constraint "CUDADriverVersion>=10.1" -attributes CUDAGlobalMemoryMb -json
+   # See which machines have that memory
```
```suggestion
   # See which machines have that much memory
```

JoshKarpel commented on docs/source/index.rst:
```diff
@@ -63,15 +70,29 @@ Happy mapping!
+   tips-and-tricks
+   faq
+
+.. toctree::
+   :maxdepth: 2
+   :hidden:
+   :caption: Use and Reference
```
Love the subsections!

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+Separate job submission/monitoring/collection
+---------------------------------------------
+
+The CLI is useful to monitor and modify ongoing jobs. Generally, in simple use
+cases the HTMap developers write three scripts:
```
```suggestion
The CLI is useful to monitor and modify ongoing jobs. Generally, in simple use
cases we recommend writing two or three scripts:
```

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
 
+.. _cli-tips:
+
+Separate job submission/monitoring/collection
```
```suggestion
Separate Job Submission/Monitoring/Collection
```

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+cases the HTMap developers write three scripts:
+
+* A script for job submission (which is run once).
+* A script for monitoring jobs (which is run many times).
```
```suggestion
* Use the CLI or a script for monitoring jobs (which is run many times).
```

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+
+* Submission: HTMap's Python API is primarily used here, possibly through
+  :func:`map`.
+* Monitoring: CLI usage is heavy here, and the command ``htmap status`` is
+  heavily used. If any of the jobs fail, it's easy to diagnose why with the
+  additional commands.
```
```suggestion
* Monitoring: CLI usage is heavy here. 
  ``htmap status`` is a good way to view a summary. 
  If any of the jobs fail, diagnose why with
  commands like ``htmap reasons`` or ``htmap errors``.
```

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+* Collection: the completed jobs are collected (as mentioned in
+  :ref:`successful-jobs`) and the results are written to disk/etc.
+
+Use of the CLI is useful for debugging when any of the scripts produces an
+error.
```
```suggestion
The CLI is useful for debugging when dealing with component holds and execution errors.
It can be used to quickly view the same kind of information as the :class:`Map` API
(though we recommend loading up the map in Python once you need to do anything
more complex than read text).
```

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+Use the CLI
+-----------
+
+This will reveal the status of each job:
```
```suggestion
This CLI command will show the status of each of your maps:
```

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+
+.. code:: shell
+
+   htmap status  # See info on each job, and various tags
```
maybe a note about `htmap status --live` as well?

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+
+   htmap status  # See info on each job, and various tags
+
+These commands will reveal more information about each job:
```
```suggestion
These commands will show more information about individual maps and map components:
```

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -4,10 +4,77 @@ Tips and Tricks
+   htmap logs  # get path to log file; info here is useful for debugging
+   htmap components foo  # view which component status for tag "foo"
+   htmap errors foo # view all errors for tag "foo"
+   htmap stdout foo 0  # view stdout for first job of tag "foo"
+   htmap stderr foo 0  # view stdout for first job of tag "foo"
```
```suggestion
   htmap stdout foo 0  # view stdout for first component of tag "foo"
   htmap stderr foo 0  # view stdout for first component of tag "foo"
```

JoshKarpel commented on docs/source/tips-and-tricks.rst:
```diff
@@ -65,10 +132,3 @@ Here's a brief example: grouping integer by whether they are even or not.
     # False [((1,), {}), ((3,), {}), ((5,), {}), ((7,), {}), ((9,), {})]
-
-
-Conditional Execution on Cluster vs. Submit
```
Did this get moved somewhere else? I have actually had a use for this. I think it's probably rarely going to be necessary, but this seems as good a place as any to stash it.

stsievert replied:
Yeah, I show it as being in the same file but moved down a bit: https://github.com/htcondor/htmap/pull/191/files#diff-715600acf23027618322ceb355f10715R74-R78

JoshKarpel commented on htmap/maps.py:
```diff
@@ -583,6 +583,34 @@ def component_statuses(self) -> List[state.ComponentStatus]:
+               sleep(x)
+               return 1 / x
+
+           futures = htmap.map(job, [0, 2, 4, 6, 8], tag="foo")
+
+           # Wait for all jobs to finish.
+           # Alternatively, use `futures = htmap.load("foo")` on a different process
+           sleep(10)
+
+           jobs = futures.components_by_status()
+           jobs = {k.name: v for k, v in jobs.items()}  # nicer key names
+           for future in jobs["COMPLETED"]:
+               result = futures.get(future)
+               # Whatever processing needs to be done
+               print(result)  # prints "2", "4", "6", and "8"
```
```suggestion
           m = htmap.map(job, [0, 2, 4, 6, 8], tag="foo")

           # Wait for all jobs to finish.
           # Alternatively, use `futures = htmap.load("foo")` on a different process
           sleep(10)

           completed = m.components_by_status()[htmap.JobStatus.COMPLETED]
           for component in completed:
               result = m.get(component)
               # Do whatever processing needs to be done
```

stsievert replied:
I found myself getting annoyed with having to type on `htmap.JobStatus.COMPLETED` when doing the work that led to this example. That's a little nicer. I have a slight preference to keep it `"COMPLETED"` as a key.

stsievert replied:
IIRC, `htmap.JobStatus.COMPLETED` doesn't much information. Is that accurate? Could the keys be changed to `"COMPLETED"` instead?

JoshKarpel replied:
Silly me, I just realized that my fancy `StrEnum` is not actually a `StrEnum`...

Just pushed https://github.com/htcondor/htmap/pull/196/commits/48d6e80dc59f99bcc06e8ddf855b265f74f10b83, you should now be able to use `"COMPLETED"` as the key without even doing the intermediate transformation (and let's do it that way in the example).

(And, of course, I typo'd and the enum is actually called `htmap.ComponentStatus`...)

stsievert replied:
Thanks! That's a lot nicer. I've modified the example. I've left in the `futures` and `jobs` variable names because I find that more clear and consistent with other Python libraries (threading, Dask, multiprocessing, etc).

JoshKarpel replied:
I think I have to insist on the variable names for consistency with the rest of the documentation. I'm open to a discussion about changing the naming scheme, but that should go on a separate issue.

JoshKarpel commented on htmap/options.py:
```diff
@@ -86,6 +86,38 @@ def __init__(
+              bbb = false
+              ccc = 1
+
+           Then, the ``MapOptions`` that would encode submit file would be:
```
```suggestion
           The ``MapOptions`` that express the same submit options would be:
```

stsievert commented on docs/source/htcondor.rst:
```diff
@@ -0,0 +1,100 @@
+Specification on exactly what "held", "idle", "running" and "completed" jobs
+mean is detailed at the `HTCondor documentation on machine states`_.
+
+.. _HTCondor documentation on machine states: https://htcondor.readthedocs.io/en/latest/admin-manual/policy-configuration.html#machine-states
```
I've made that change, and also have added note that says this:

> For more detail, see the relevant HTCondor documentation:
>
> * https://htcondor.readthedocs.io/en/latest/users-manual/managing-a-job.html#checking-on-the-progress-of-jobs
> * https://htcondor.readthedocs.io/en/latest/admin-manual/policy-configuration.html#machine-states

JoshKarpel commented:
Looks like everything is resolved! I'll let it simmer overnight for any final changes and merge it tomorrow.

(Don't worry about the CI builds... there's a flaky test that I've been trying to hammer out.)

stsievert commented:
I'll try to give it another review tonight too.

JoshKarpel commented:
Thank you so much @stsievert !
